include/zephyr/drivers/uart.h

```diff
@@ -1160,22 +1160,30 @@ static inline int z_impl_uart_irq_update(const struct device *dev)
  * @param dev UART device instance.
  * @param cb Pointer to the callback function.
  * @param user_data Data to pass to callback function.
+ *
+ * @retval 1 On success.
+ * @retval -ENOSYS If this function is not implemented.
+ * @retval -ENOTSUP If API is not enabled.
  */
-static inline void uart_irq_callback_user_data_set(const struct device *dev,
-						   uart_irq_callback_user_data_t cb,
-						   void *user_data)
+static inline int uart_irq_callback_user_data_set(const struct device *dev,
+						  uart_irq_callback_user_data_t cb,
+						  void *user_data)
 {
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 	const struct uart_driver_api *api =
 		(const struct uart_driver_api *)dev->api;
 
 	if ((api != NULL) && (api->irq_callback_set != NULL)) {
 		api->irq_callback_set(dev, cb, user_data);
+		return 1;
+	} else {
+		return -ENOSYS;
 	}
 #else
 	ARG_UNUSED(dev);
 	ARG_UNUSED(cb);
 	ARG_UNUSED(user_data);
+	return -ENOTSUP;
```

Comment:
If we strictly follow the guidelines for [return codes](https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes) this is not the correct return code, as this may be supported by the driver. However, I see it's used in the same way in the entire UART driver, so it should be fine.

Reply:
I don't find those guidelines very clear to be honest.  The author seems to have very precise definitions for "implemented" and "supported", but I'm not sure what they are.  It's also not clear how the Zephyr build time configuration is supposed to be reflected.  That's a very real distinction that actually matters, since it tells the user they made a configuration mistake that can be fixed.

I guess it was also not clear to others as well, since the author of the UART API seemed to have a different idea how to describe config choices than you do.